0000-access-keys.md

```diff
@@ -0,0 +1,313 @@
+- Proposal Code Name: access_keys
+- Start Date: 2019-07-08
+- NEP PR: [nearprotocol/neps#0000](https://github.com/nearprotocol/neps/pull/0000)
+- Issue(s): [nearprotocol/nearcore#687](https://github.com/nearprotocol/nearcore/issues/687)
+
+# Summary
+[summary]: #summary
+
+Access keys provide limited access to an account.
+Each access key belongs to some account and identified by a unique (within the account) public key.
+One account may have large number of access keys.
+Access keys will replace original account-level public keys.
+Access keys allow to act on behalf of the account by restricting allowed transactions with the access key permissions.
+
+# Motivation
+[motivation]: #motivation
+
+Access keys give an ability to use dApps in a secure way without asking the user to sign every transaction in the wallet.
+By issuing the access key once for the application, the application can now act on behalf of the user in a restricted environment.
+This enables seamless experience for the user.
+
+Access keys also enable a few other use-cases that are discussed in details below.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Here are proposed changes for the AccessKey and Account structs.  
+
+```rust
+/// `account_id,public_key` is a key in the state
+struct AccessKey {
+  /// The nonce for this.
+  pub nonce: Nonce,  // u64 
+  
+  /// Defines permissions for the AccessKey 
+  pub permission: AccessKeyPermission,
+}
+
+/// Defines permissions for AccessKey 
+pub enum AccessKeyPermission {
+  /// Restricts AccessKey to only be used for function calls.
+  FunctionCall(FunctionCallPermission),
+
+  /// Gives full access to the account.
+  /// NOTE: It's used to replace account-level public keys.
+  FullAccess,
+}
+
+pub struct FunctionCallPermission {
+  /// The amount that can be spent for transaction fees by this access key from the account balance.
+  /// When used, both account balance and the allowance is decreased.
+  pub allowance: Balance,  // u128
```

Comment:
Idea: if allowance is equal to max u128 value then we do not decrement.

Reply:
The better option would be what @bowenwang1996 suggested with using Option<Balance> type (where None means unlimited allowance), but I can't think of an example when we actually need to have u128::max value instead of a full permission key.

Reply:
Using Option is more type-safe. I think it is a good idea.